Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.mm

```diff
@@ -0,0 +1,124 @@
+    {
+        if (IsEnabled())
+        {
+            STEP_COUNTER_LOG(@"Was already enabled?");
+            if (m_Device != deviceId)
+                STEP_COUNTER_LOG(@"Enabling with different device id? Expected %d, was %d. Are you creating more than one iOSStepCounter", m_Device, deviceId);
+        }
+        m_Pedometer = [[CMPedometer alloc]init];
+        m_Callbacks = *callbacks;
+        m_Device = deviceId;
+        [m_Pedometer startPedometerUpdatesFromDate: [NSDate date] withHandler:^(CMPedometerData * _Nullable pedometerData, NSError * _Nullable error)
+        {
+            // Note: We need to call our callback on the same thread the Unity scripting is operating
+            dispatch_async(dispatch_get_main_queue(), ^{
+                // Guard against situation where device was disabled, any event which are received after that, should be ignored.
+                if (m_Device == -1)
+                    return;
+                m_Callbacks.dataReceived(m_Device, [pedometerData.numberOfSteps intValue]);
+            });
+        }];
+    }
+
+    void Disable(int deviceId)
+    {
+        if (m_Pedometer == nullptr)
+            return;
+        if (m_Device != deviceId)
+            STEP_COUNTER_LOG(@"Disabling with wrong device id, expected %d, was %d", m_Device, deviceId);
+        [m_Pedometer stopPedometerUpdates];
+        m_Pedometer = nullptr;
+        m_Device = -1;
+    }
+
+    bool IsEnabled() const
+    {
+        return m_Pedometer != nullptr;
+    }
+
+private:
+    CMPedometer* m_Pedometer;
+    iOSStepCounterCallbacks m_Callbacks;
+    int m_Device;
+};
+
+static iOSStepCounterWrapper s_Wrapper;
+static const int kResultSuccess = 1;
+static const int kResultFailure = -1;
+
+extern "C" int _iOSStepCounterIsAvailable()
+{
+    return [CMPedometer isStepCountingAvailable] ? 1 : 0;
+}
+
+extern "C" int _iOSStepCounterIsAuthorized()
+{
+    if (@available(iOS 11.0, *))
+    {
+        return [CMPedometer authorizationStatus] == CMAuthorizationStatusAuthorized ? 1 : 0;
+    }
+    return 1;
+}
+
+extern "C" int _iOSStepCounterEnable(int deviceId, iOSStepCounterWrapper::iOSStepCounterCallbacks* callbacks, int sizeOfCallbacks)
+{
+    if (sizeof(iOSStepCounterWrapper::iOSStepCounterCallbacks) != sizeOfCallbacks)
+    {
+        STEP_COUNTER_LOG(@"iOSStepCounterCallbacks size mismatch, expected %lu was %d", sizeof(iOSStepCounterWrapper::iOSStepCounterCallbacks), sizeOfCallbacks);
+        return kResultFailure;
+    }
+
+    if (_iOSStepCounterIsAvailable() == 0)
+    {
+        STEP_COUNTER_LOG(@"Step counting is not available");
+        return kResultFailure;
+    }
+
+    if (_iOSStepCounterIsAuthorized() == 0)
+    {
+        STEP_COUNTER_LOG(@"Step counting was not authorized");
+        return kResultFailure;
+    }
+
+    s_Wrapper.Enable(deviceId, callbacks);
+
+    return kResultSuccess;
+}
+
+extern "C" int _iOSStepCounterDisable(int deviceId)
+{
+    s_Wrapper.Disable(deviceId);
```

Comment:
is it worth to return kResultFailure in case deviceId is diferent or m_Pedometer is null in s_Wrapper ?

Reply:
Makes sense